File: engine-desktop/src/com/magma/engine/debug/Debugger.java
package com.magma.engine.debug;

import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.BoxLayout;
import javax.swing.JFrame;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input.Keys;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Disposable;
import com.magma.engine.MagmaGame;
import com.magma.engine.debug.modules.InfoModule;
import com.magma.engine.debug.modules.MapModule;
import com.magma.engine.debug.modules.Module;
import com.magma.engine.debug.modules.StageModule;

public class Debugger extends JFrame implements Disposable {

    private static Debugger instance;

    private Array<Module> modules = new Array<Module>();

    private Debugger() {
        super("Debugger");
        instance = this; // prevents stack overflow do not remove

        // prevents the game from hanging when closing, due to the frame still being
        // available
        MagmaGame.disposeOnExit(this);

        // Construct the window
        setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        setResizable(true);
        setLayout(new BoxLayout(getContentPane(), BoxLayout.PAGE_AXIS));
        setLocationRelativeTo(null);

        final Dimension SIZE = new Dimension(400, 700);
        setMinimumSize(SIZE);
        setPreferredSize(SIZE);

        // add menubar
        JMenuBar menuBar = new JMenuBar();
        JMenu menuFile = new JMenu("File");
        JMenuItem menuItemExit = new JMenuItem("Exit");
        menuItemExit.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent arg0) {
                instance.setVisible(false);
            }
        });
        menuFile.add(menuItemExit);
        menuBar.add(menuFile);

        // adds menu bar to the frame
        setJMenuBar(menuBar);

        // generate modules
        addModule(new InfoModule());
        addModule(new MapModule());
        addModule(new StageModule());

        pack();
    }

    public static void addModule(Module module) {
        Debugger ins = getInstance();
        ins.modules.add(module);
        ins.getContentPane().add(module);
        ins.pack();
    }

    public static void update() {
        Debugger ins = getInstance();
        if (Gdx.input.isKeyJustPressed(Keys.F3)) {
            ins.setVisible(!getInstance().isVisible());
        }

        if (ins.isVisible()) {
            for (Module module : ins.modules) {
                module.update();
            }
        }
    }

    public static <T extends Module> Module getModule(Class<T> type) {
        for (Module mod : getInstance().modules) {
            MagmaLogger.log(type);
            if (mod.getClass() == type) {
                return mod;
            }
        }
        throw new NullPointerException("Cannot find module " + type);
    }

    public static Debugger getInstance() {
        if (instance == null) {
            instance = new Debugger();
        }
        return instance;
    }

    @Override
    public void dispose() {
        modules = null;
        super.dispose();
        MagmaLogger.log(this, "Debugger disposed");
    }
}

File: engine/src/com/magma/engine/debug/Debugger.kt
package com.magma.engine.debug

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.utils.Disposable
import com.magma.engine.MagmaGame
import com.magma.engine.debug.modules.Module
import java.awt.Dimension
import java.lang.IllegalStateException
import javax.swing.*

class Debugger private constructor(): JFrame("Debugger"), Disposable {

    init {
        // prevents the game from hanging when closing, due to the frame still being
        // available
        MagmaGame.disposeOnExit(this)

        // Construct the window
        defaultCloseOperation = DISPOSE_ON_CLOSE
        isResizable = true
        layout = BoxLayout(contentPane, BoxLayout.PAGE_AXIS)
        setLocationRelativeTo(null)
        val size = Dimension(400, 700)
        minimumSize = size
        preferredSize = size

        // add menubar
        val menuBar = JMenuBar()
        val menuFile = JMenu("File")
        val menuItemExit = JMenuItem("Exit")
        menuItemExit.addActionListener { isVisible = false }
        menuFile.add(menuItemExit)
        menuBar.add(menuFile)

        // adds menu bar to the frame
        jMenuBar = menuBar

        pack()
    }

    override fun dispose() {
        super.dispose()
        MagmaLogger.log(this, "Debugger disposed")
    }

    companion object {
        private val modules: HashMap<Any,Module> = HashMap()
        private var instance: Debugger? = null

        var enabled: Boolean = false
            set(value) {
                field = value
                MagmaLogger.log(this,"Debug mode is "+if (field) "on" else "off")
            }

        fun addModule(owner: Any,module: Module) {
            if (!enabled) return

            if (instance == null){
                // lazy init the debugger
                instance = Debugger()
            }

            val prev = modules.put(owner,module)
            if (prev != null) throw IllegalStateException("Duplicate module types")

            instance!!.contentPane?.add(module)
            instance!!.pack()
        }

        fun removeOwnedModule(owner: Any){
            modules.remove(owner)
        }

        fun update() {
            if (Gdx.input.isKeyJustPressed(Input.Keys.F3)) {
                if (enabled && instance == null){
                    instance = Debugger()
                }
                instance?.isVisible = !instance!!.isVisible
            }
            if (instance != null && instance!!.isVisible) {
                for (module in modules.values) {
                    module.update()
                }
            }
        }
    }
}
